### PR TITLE
Fix empty space skopt

### LINKED ISSUE #77: Issue with skopt optimizer
Evaluate failed.
Traceback (most recent call last):
  File "/var/src/auger-ml-worker/auger_ml/auger_ml/optimizers/skopt_base_optimizer.py", line 33, in evaluate
    outstanding_trials = [self.update_trial(self._suggest()) for _ in range(self.batch_size)]
  File "/var/src/auger-ml-worker/auger_ml/auger_ml/optimizers/skopt_base_optimizer.py", line 50, in _suggest
    alg = np.random.choice(list(self.space.keys()))
  File "mtrand.pyx", line 1126, in mtrand.RandomState.choice
ValueError: a must be non-empty

### PR DESCRIPTION
**Pull request by jetsnguns**
Fix #77

### CODE REVIEW

File: auger_ml/optimizers/hyperopt_base_optimizer.py [hunk 1]
```diff
@@ -27,7 +28,8 @@ def __init__(self, optimizer_params, options, trialsQueue, optimizersQueue):
         def fn_for_hp_domain(x):
             return {"loss": 0.0, "status": hp_base.STATUS_OK, "cfg": x}
 
-        self.hp_domain = hp_base.Domain(fn_for_hp_domain, self._convert_auger_space_to_hp(self.options['search_space']),
+        self.trimmed_space = trim_auger_space(self.options['search_space'])
+        self.hp_domain = hp_base.Domain(fn_for_hp_domain, self._convert_auger_space_to_hp(self.trimmed_space),
                                         pass_expr_memo_ctrl=False)
         self.alg_list = [alg for alg, _ in self.options['search_space'].items()]
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/hyperopt_base_optimizer.py]: 1. As far as I understand self.alg_list is equal to self.options['search_space'].keys().
2. As far as I understand self.options['search_space'].keys() is equal to self.trimmed_space.keys()
If both point above are true please replace the statement with self.alg_list = self.trimmed_space.keys()

File: auger_ml/optimizers/hyperopt_base_optimizer.py [hunk 2]
```diff
@@ -154,17 +159,9 @@ def _convert_parameter_to_hp(alg, name, param_dic):
...
-            if p['type'] in ['float', 'int']:
-                low = p['bounds'][0]
-                high = p['bounds'][1]
-                if high <= low:
-                    return False
-            return True
-
         domain = hp.choice("ml_algorithm_dict",
                            [{"algorithm_name": alg,
                              "algorithm_params": {name: HyperoptBaseOptimizer._convert_parameter_to_hp(alg, name, pd)
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/hyperopt_base_optimizer.py]: It's better to write cls._convert_parameter_to_hp instead HyperoptBaseOptimizer._convert_parameter_to_hp.

File: auger_ml/optimizers/hyperopt_base_optimizer.py [hunk 3]
```diff
@@ -154,17 +159,9 @@ def _convert_parameter_to_hp(alg, name, param_dic):
 
     @staticmethod
     def _convert_auger_space_to_hp(search_space):
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/hyperopt_base_optimizer.py]: Make _convert_auger_space_to_hp classmethod.

File: auger_ml/optimizers/hyperopt_base_optimizer.py [hunk 4]
```diff
@@ -81,20 +83,23 @@ def auger_trial_to_hp(self, auger_trial, new_id):
...
-                for result in results:
-                    if not result.get("error", None):
-                        scores.append(result['result'])
-        except Exception as e:
-            if self.options.get('pass_exception', False):
-                raise
-
-            if not isinstance(e, TasksException):
-                logger.exception("Evaluate failed.")
+        if self.trimmed_space:
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/hyperopt_base_optimizer.py]: Please explain why do you guard code with this if statement.

File: auger_ml/optimizers/skopt_base_optimizer.py [hunk 5]
```diff
@@ -20,28 +21,32 @@ def __init__(self, optimizer_params, options, trialsQueue, optimizersQueue):
...
-                for result in results:
-                    if not result.get("error", None):
-                        scores.append(result['result'])
-        except Exception as e:
-            if self.options.get('pass_exception', False):
-                raise
-
-            if not isinstance(e, TasksException):
-                logger.exception("Evaluate failed.")
+        if self.trimmed_space:
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/skopt_base_optimizer.py]: Please explain why do you guard code with this if statement.

File: auger_ml/optimizers/space/base.py [hunk 6]
```diff
@@ -12,4 +12,33 @@
...
+        return True
+
+    def trim_subspace(p):
+        tp = {n: d for n, d in p.items() if check_param(d)}
+        return tp
+
+    if not auger_space:
+        return True
+
+    checked_space = {n: trim_subspace(p) for n, p in auger_space.items()}
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/space/base.py]: checked_space could be converted to generator of tuples of kind (n, trim_subspace(p)).

File: auger_ml/optimizers/space/base.py [hunk 7]
```diff
@@ -12,4 +12,33 @@
...
+
+    trimmed_space = {n: p for n, p in checked_space.items() if p}
```
Comment (StanislavVolodarskiy) [inline auger_ml/optimizers/space/base.py]: Here we drop algo if there are no parameters to tune. But algo without parameters can show better results then others. So why we drop them? If you have a reason please add comment.

**Comment by jetsnguns** [comment]
Added requested changes.

**Pull request merged**